modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/memory/internal/ExternalIdIndexUtils.kt

```diff
@@ -0,0 +1,67 @@
+package com.r3.corda.lib.tokens.selection.memory.internal
+
+import net.corda.core.crypto.toStringShort
+import net.corda.core.node.ServiceHub
+import java.security.PublicKey
+import java.util.*
+
+/** Table names. */
+val publicKeyHashToExternalId = "pk_hash_to_ext_id_map"
+
+/** Column names. */
+val publicKeyHashToExternalId_externalId = "external_id"
+val publicKeyHashToExternalId_publicKeyHash = "public_key_hash"
+
+sealed class Holder {
+    data class KeyIdentity(val owningKey: PublicKey) : Holder() // Just public key
+    object UnmappedIdentity : Holder() // For all keys that are unmapped
+    data class MappedIdentity(val uuid: UUID) : Holder() // All keys register to this uuid
+    object TokenOnly : Holder() // This is for the case where we use token class and token identifier only
+
+    companion object {
+        fun fromUUID(uuid: UUID?): Holder {
+            return if (uuid != null) {
+                MappedIdentity(uuid)
+            } else {
+                UnmappedIdentity
+            }
+        }
+    }
+}
+
+// TODO after corda 4.3 we will be able to use cache introduced by this PR https://github.com/corda/corda/pull/5357
+fun lookupExternalIdFromKey(owningKey: PublicKey, serviceHub: ServiceHub): Holder {
```

Comment:
Once RC01 is out for 4.3 then you can rely on the identity service for doing this.